pkg/alertmanager/distributor_test.go

```diff
@@ -117,128 +124,142 @@ func TestDistributor_DistributeRequest(t *testing.T) {
+			name:              "Read /v2/silence/id is sent to 3 AMs",
+			numAM:             5,
+			numHappyAM:        5,
+			replicationFactor: 3,
+			isRead:            true,
+			expStatusCode:     http.StatusOK,
+			expectedMinCalls:  3,
+			expectedMaxCalls:  3,
+			route:             "/v2/silence/id",
+			responseBody:      []byte(`{"id":"aaa","updatedAt":"2020-01-01T00:00:00Z"}`),
 		},
 		{
 			name:                "Write /silence/id not supported",
 			numAM:               5,
 			numHappyAM:          5,
 			replicationFactor:   3,
 			expStatusCode:       http.StatusNotFound,
-			expectedTotalCalls:  0,
+			expectedMinCalls:    0,
+			expectedMaxCalls:    0,
 			headersNotPreserved: true,
 			route:               "/silence/id",
 		}, {
-			name:               "Delete /silence/id is sent to only 1 AM",
-			numAM:              5,
-			numHappyAM:         5,
-			replicationFactor:  3,
-			isDelete:           true,
-			expStatusCode:      http.StatusOK,
-			expectedTotalCalls: 1,
-			route:              "/silence/id",
+			name:              "Delete /silence/id is sent to only 1 AM",
+			numAM:             5,
+			numHappyAM:        5,
+			replicationFactor: 3,
+			isDelete:          true,
+			expStatusCode:     http.StatusOK,
+			expectedMinCalls:  1,
+			expectedMaxCalls:  1,
+			route:             "/silence/id",
 		}, {
-			name:               "Read /status is sent to only 1 AM",
-			numAM:              5,
-			numHappyAM:         5,
-			replicationFactor:  3,
-			isRead:             true,
-			expStatusCode:      http.StatusOK,
-			expectedTotalCalls: 1,
-			route:              "/status",
+			name:              "Read /status is sent to only 1 AM",
+			numAM:             5,
+			numHappyAM:        5,
+			replicationFactor: 3,
+			isRead:            true,
+			expStatusCode:     http.StatusOK,
+			expectedMinCalls:  1,
+			expectedMaxCalls:  1,
+			route:             "/status",
 		}, {
-			name:               "Read /status should try all alert managers on error",
-			numAM:              3,
-			numHappyAM:         0,
-			replicationFactor:  3,
-			isRead:             true,
-			expStatusCode:      http.StatusInternalServerError,
-			expectedTotalCalls: 3,
-			route:              "/status",
+			name:              "Read /status should try all alert managers on error",
+			numAM:             3,
+			numHappyAM:        0,
+			replicationFactor: 3,
+			isRead:            true,
+			expStatusCode:     http.StatusInternalServerError,
+			expectedMinCalls:  3,
+			expectedMaxCalls:  3,
+			route:             "/status",
 		}, {
-			name:               "Read /status is sent to 3 AM when 2 are not happy",
-			numAM:              3,
-			numHappyAM:         1,
-			replicationFactor:  3,
-			isRead:             true,
-			expStatusCode:      http.StatusOK,
-			expectedTotalCalls: 3,
-			route:              "/status",
+			name:              "Read /status is sent to atleast 2 AM when 2 are not happy",
```

Comment:
This test was failing because the totalCalls != `expectedTotalCalls`
This could happen because of the order in which the alert-manager instances are shuffled.
Since we return when we get the first success response from 1 AM, we don't have to always call 3 AMs.
I changed the test to make sure that `2 < = total calls <= 3`

Reply:
I hate to be picky. but do you mind creating another PR to fix this test case. It's unrelated to this PR

Reply:
I did not see this PR.. I've already fixed it.
https://github.com/cortexproject/cortex/pull/4865

Reply:
Rebased Alan's changes